lls/nd.c

```diff
@@ -0,0 +1,736 @@
+
+		icmpv6_hdr->cksum = rte_ipv6_icmpv6_cksum(ipv6_hdr, icmpv6_hdr);
+
+		if (rte_eth_tx_burst(iface->id, tx_queue, &buf, 1) <= 0) {
+			RTE_LOG(ERR, GATEKEEPER, "lls: could not send an ND Neighbor Solicitation in response to a Solicitation\n");
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * RFC 4861, Section 7.2.5: Receipt of Neighbor Advertisements.
+ *
+ * The RFC states that there is no need to create an entry if
+ * none exists, but we do.
+ *
+ * We do not adhere to the meanings of the Router, Solicited,
+ * or Override flags when it comes to updating an entry (although
+ * we do use the Solicited flag in a validity test), because
+ * we make no distinction between cache entry states such as
+ * STALE, INCOMPLETE, and UNREACHABLE as described by the RFC.
+ * We don't care about whether an entry is a router or whether an
+ * announcement was solicited, we do not implement Neighbor
+ * Unreachability Detection, and we always update an entry
+ * even when the Override flag is not set.
+ */
+static int
+process_nd_neigh_advertisement(struct lls_config *lls_conf,
+	struct ipv6_hdr *ipv6_hdr, struct icmpv6_hdr *icmpv6_hdr,
+	uint16_t icmpv6_len, struct gatekeeper_if *iface)
+{
+	struct nd_neigh_msg *nd_msg = (struct nd_neigh_msg *)&icmpv6_hdr[1];
+	struct nd_opt_lladdr *nd_opt;
+	struct nd_opts ndopts;
+
+	/*
+	 * Most of the checks required by RFC 4861, Section 7.1.2
+	 * have already been done by nd_pkt_valid().
+	 */
+
+	/*
+	 * RFC 4861, Section 7.1.2.
+	 *
+	 * Target Address must not be a multicast address.
+	 */
+	if (ipv6_addr_multicast(nd_msg->target))
+		return -1;
+
+	/*
+	 * RFC 4861, Section 7.1.2.
+	 *
+	 * If the IP Destination Address is a multicast address
+	 * the Solicited flag must be zero.
+	 */
+	if (ipv6_addr_multicast(ipv6_hdr->dst_addr) &&
+			(nd_msg->flags & rte_cpu_to_be_32(LLS_ND_NA_SOLICITED)))
+		return -1;
+
+	/* Process any ND neighbor options and save them in @ndopts. */
+	if (parse_nd_opts(&ndopts, nd_msg->opts, icmpv6_len -
+			(sizeof(*icmpv6_hdr) + sizeof(*nd_msg))) == NULL)
+		return -1;
+
+	if (ndopts.opt_array[ND_OPT_TARGET_LL_ADDR] != NULL) {
+		struct lls_mod_req mod_req = {
+			.cache = &lls_conf->nd_cache,
+			.port_id = iface->id,
+			.ts = time(NULL),
+		};
+
+		RTE_VERIFY(mod_req.ts >= 0);
+		nd_opt = (struct nd_opt_lladdr *)
+			ndopts.opt_array[ND_OPT_TARGET_LL_ADDR];
+
+		rte_memcpy(mod_req.ip_be, nd_msg->target,
+			lls_conf->nd_cache.key_len);
+		ether_addr_copy(&nd_opt->ha, &mod_req.ha);
+		lls_process_mod(lls_conf, &mod_req);
+	}
+
+	/* Always need to free the packet. */
+	return -1;
+}
+
+/*
+ * Perform sanity checks to make sure this is a valid ND neighbor packet.
+ * By RFC 4861, sections 7.1.1 and 7.1.2, these checks are required for both
+ * Solicitations and Advertisements.
+ */
+static int
+nd_pkt_valid(struct ipv6_hdr *ipv6_hdr, struct icmpv6_hdr *icmpv6_hdr,
+	uint16_t icmpv6_len)
+{
+	uint16_t cksum; /* Need variable to avoid compiler error. */
```

Comment:
Here's the message:

/home/cody/gatekeeper/lls/nd.c:671:48: error: promoted ~unsigned is always non-zero [-Werror=sign-compare]
   ~rte_ipv6_icmpv6_cksum(ipv6_hdr, icmpv6_hdr) == 0;
                                                ^
compilation terminated due to -Wfatal-errors.
cc1: all warnings being treated as errors